pkg/ingester/client/circuitbreaker.go

```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package client
+
+import (
+	"context"
+	"errors"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/sony/gobreaker"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+)
+
+const (
+	resultSuccess = "success"
+	resultError   = "error"
+	resultOpen    = "circuit_breaker_open"
+)
+
+func NewCircuitBreaker(name string, cfg CircuitBreakerConfig, metrics *Metrics, logger log.Logger) grpc.UnaryClientInterceptor {
+	breaker := gobreaker.NewCircuitBreaker(gobreaker.Settings{
+		Name:        name,
+		Interval:    cfg.ClosedInterval,
+		Timeout:     cfg.OpenTimeout,
+		MaxRequests: uint32(cfg.MaxHalfOpenRequests),
+		ReadyToTrip: func(counts gobreaker.Counts) bool {
+			return uint64(counts.ConsecutiveFailures) >= cfg.MaxConsecutiveFailures
+		},
+		OnStateChange: func(name string, from gobreaker.State, to gobreaker.State) {
+			level.Info(logger).Log("msg", "circuit breaker changing state", "name", name, "from", from, "to", to)
```

Comment:
These logs don't have context on which circuit breaker it is. I know we only have 1, but when reading the logs there's nothing saying "ingester client circuit breaker". What do you think if you wrap the logger before passing it to the circuit breaker, adding a field like `"component", "ingester client"`.

Reply:
Fixed in e9fc92026c936af1eb2d9d83aad333f1a31ddbda